### PR TITLE
DROOLS-2759: [DMN Editor] Custom data-types - Create the treegrid-table component

### PR DESCRIPTION
**Pull request by karreiro**
See:
- https://issues.jboss.org/browse/DROOLS-2759
- https://issues.jboss.org/browse/DROOLS-2760
- https://issues.jboss.org/browse/DROOLS-2702
- https://issues.jboss.org/browse/DROOLS-2761
- https://issues.jboss.org/browse/DROOLS-2717

drools-2759-demo

---

1. The goal of this PR is to cover the treegrid-table component just in read-only mode.

2. This [DMN file](https://gist.github.com/karreiro/d84a51bc666123542499b96061c13956) has some data types that can be used for the tests.

### CODE REVIEW

**Comment by jomarko** [comment]
@karreiro for now I just measured mutation coverage and prepared [PR](https://github.com/karreiro/kie-wb-common/pull/3) for it. I still have to try manually and have a look on all code.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/DataType.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+    public String getType() {
+        return type;
+    }
+
+    public List<DataType> getSubDataTypes() {
+        return subDataTypes;
+    }
+
+    public boolean isBasic() {
```
Comment (manstis) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/DataType.java]: Can you explain "basic" vs "default" vs "external"? DMN terminology is "built in" (for standard _out of the box_ data-types) and "user/custom" (defined by the User) and "imported" (defined in another DMN file and imported by the User - we do not support _imports_ at the moment in our editor).

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/DataTypeModal.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        super.setup();
+        getView().setup(treeGrid);
+    }
+
+    public void show(final String value) {
+
+        final String dataTypeName = extractItemDefinitionName(value);
+        final Optional<ItemDefinition> itemDefinition = itemDefinitionUtils.findByName(dataTypeName);
+
+        if (itemDefinition.isPresent()) {
```
Comment (manstis) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/DataTypeModal.java]: I know @jomarko is going to say you should use ```itemDefinition.ifPresent(id->{...})```!

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/common/DataTypeFactory.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    private String extractType(final ItemDefinition itemDefinition) {
+        final Optional<QName> typeRef = Optional.ofNullable(itemDefinition.getTypeRef());
+        return typeRef.isPresent() ? typeRef.get().getLocalPart() : structure();
+    }
+
+    private boolean isBasic(final ItemDefinition itemDefinition) {
+        return Optional.ofNullable(itemDefinition.getTypeRef()).isPresent();
+    }
+
+    public boolean isDefaultDefault(final String type) {
```
Comment (manstis) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/common/DataTypeFactory.java]: ```isDefaultDefault(..)```? Why the double ```default```?

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/common/DataTypeFactory.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        return name.isPresent() ? name.get().getValue() : none();
+    }
+
+    private String extractType(final ItemDefinition itemDefinition) {
+        final Optional<QName> typeRef = Optional.ofNullable(itemDefinition.getTypeRef());
+        return typeRef.isPresent() ? typeRef.get().getLocalPart() : structure();
+    }
+
+    private boolean isBasic(final ItemDefinition itemDefinition) {
+        return Optional.ofNullable(itemDefinition.getTypeRef()).isPresent();
```
Comment (manstis) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/common/DataTypeFactory.java]: It what scenarios does an ```ItemDefinition``` not have a ```TypeRef```?! The DMN v1.1 Specification (7.3.2 ItemDefinition metamodel) states "An ItemDefinition element SHALL have a typeRef" therefore would not all data-types be considered "basic" by this code?

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/common/DataTypeFactory.java (hunk 3 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/common/DataTypeFactory.java]: Considering it looks for a match with a "built in type" is use of (standard) DMN terminology ```isBuiltin(..)``` not better?

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/common/DataTypeUtils.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    private final DMNGraphUtils dmnGraphUtils;
+
+    @Inject
+    public DataTypeUtils(final DataTypeFactory dataTypeFactory,
+                         final DMNGraphUtils dmnGraphUtils) {
+        this.dataTypeFactory = dataTypeFactory;
+        this.dmnGraphUtils = dmnGraphUtils;
+    }
+
+    public List<DataType> defaultDataTypes() {
```
Comment (manstis) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/common/DataTypeUtils.java]: Can we use "built in" vs "default"?

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeSelect.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    public void init(final DataType dataType) {
+        this.dataType = dataType;
+        this.view.setDataType(dataType);
+    }
+
+    DataType getDataType() {
+        return dataType;
+    }
+
+    List<DataType> getDefaultDataTypes() {
```
Comment (manstis) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeSelect.java]: Sorry, can we use "built in" again please?

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeSelectView.java [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+    void setupDropdown() {
+        setupDropdownItems();
+        setupSelectPicker();
+        hideSelectPicker();
+        setupSelectPickerOnChangeHandler();
+    }
+
+    void setupDropdownItems() {
+        typeSelect.innerHTML = "";
+        typeSelect.appendChild(makeOptionGroup(defaultTitle(), presenter.getDefaultDataTypes(), DataType::getType));
```
Comment (manstis) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeSelectView.java]: Did you want me to call out every "default" vs "built in"?

I'll stop from here on.. but let me know if you'd prefer me to search.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeSelectView.java [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+        show(typeText);
+        hideSelectPicker();
+    }
+
+    void setPickerValue(final String value) {
+        setPickerValue(getSelectPicker(), value);
+    }
+
+    HTMLOptionElement makeHTMLOptionElement() {
+        // This is a workaround for an issue on Errai (related to 'ManagedInstance' + 'HTMLOptionElement').
```
Comment (manstis) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeSelectView.java]: That must have been a nice bug to track down!

Is there a Errai JIRA for it (is it worth linking the JIRA, who knows one day someone else may look at this!)

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeSelectView.java [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+
+    void openSelectPicker() {
+        triggerPickerAction(getSelectPicker(), "toggle");
+    }
+
+    Element getSelectPicker() {
+        return getElement().querySelector("[data-field='type-select']");
+    }
+
+    native void triggerPickerAction(final Element element,
```
Comment (manstis) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeSelectView.java]: These are interesting...

In ideal world we should probably look to use ```@JsInterop``` vs ```JSNI``` for these interactions.

That said IDK if it's possible without a dis-proportionate amount of effort. Paulo may have some thoughts.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeTreeGridItem.java [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+
+    DataType getDataType() {
+        return dataType;
+    }
+
+    public int getLevel() {
+        return level;
+    }
+
+    void expandOrCollapseSubTypes() {
```
Comment (manstis) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeTreeGridItem.java]: Arguably this _presenter_ has too much knowledge of the _view_ (as it uses both ```HTMLElement``` and manipulates CSS on the _view_ components). Did you want to consider improving the separation (or convince me this is OK!) ;-)

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeTreeGridItemView.java [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+    void setupArrow(final DataType dataType) {
+        if (dataType.hasSubDataTypes()) {
+            show(arrow);
+        } else {
+            hide(arrow);
+        }
+    }
+
+    void setupIndentationLevel() {
+        level.style.marginLeft = margin(20 * presenter.getLevel());
```
Comment (manstis) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeTreeGridItemView.java]: Care to move the magic number to a constant?

**Comment by karreiro** [comment]
@manstis Thank you for the review.

Regarding the terminology "Built in" vs. "Default", as discussed in Slack, let's keep the UI code following the UI prototypes (uncoupled from the DMN spec). Let's review all terms, considering the following prototype:

prototype

- Basic - is a regular data type, with a name and a type. Example: "name (Text)", "age (Number)" and "Address (tAddress)"
- External - is a data type that cannot be updated, because it belongs to an existing (external) type. Example: "street (Text)", "city (Text)" and "postal code (Text)".
- Default - is a primitive data type. Example: "name (Text)", "age (Number)"

Notice:
- "Employer" is not a basic data type, it's a structure/nested data type;
- "Address" is basic because it _has_ a type called "tAddress" (..and we can see its sub data types "street", "city" and "postal code");

---

@jomarko I incorporated your changes in this PR, except the following ones:

1) DataTypeTreeGrid.java:65 / DataTypeTreeGridTest.java - the subDataTypes variable is never `null`. Additionally, changing the code
from:
```
!subDataTypes.isEmpty()
```
to
```
subDataTypes != null
```
changes the behaviour of the method.

2) TypePickerWidgetTest.java - the changes in my PR are not related to the changes in my PR. We can incorporate your improvements, for sure, but they are not related to this PR.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/DataType.java (hunk 1 above)
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/DataType.java]: Answered here: https://github.com/kiegroup/kie-wb-common/pull/2011#issuecomment-409683570 :-)

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/DataTypeModal.java (hunk 2 above)
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/DataTypeModal.java]: Fixed :-)

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/common/DataTypeFactory.java (hunk 3 above)
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/common/DataTypeFactory.java]: Answered here: https://github.com/kiegroup/kie-wb-common/pull/2011#issuecomment-409683570 :-)

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/common/DataTypeFactory.java (hunk 4 above)
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/common/DataTypeFactory.java]: The `typeRef` is `null` when the `ItemDefinition` has sub items (`itemComponent`) (like the Employer), here:
prototype

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeSelect.java (hunk 6 above)
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeSelect.java]: Answered here: https://github.com/kiegroup/kie-wb-common/pull/2011#issuecomment-409683570 :-)

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeSelectView.java (hunk 7 above)
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeSelectView.java]: Hehe - answered here: https://github.com/kiegroup/kie-wb-common/pull/2011#issuecomment-409683570 :-)

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeSelectView.java (hunk 8 above)
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeSelectView.java]: I raised the JIRA [here](https://issues.jboss.org/browse/ERRAI-1114) ;-)

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeTreeGridItem.java (hunk 10 above)
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeTreeGridItem.java]: My first approach implemented this logic in the View. However, considering the MVP pattern, I see the View as a passive element and the Presenter as something that consults View values and performs operations to change it considering the Model.

In this scenario, the Presenter knows the View a little bit more (since it has some query selectors there). But, I preferred to do this, instead of creating an indirection and ask for the `DataTypeTreeGrid` execute an operation on specifics Views. Here I'm keeping the `DataTypeTreeGridItem` uncoupled, and I'm using the DOM tree to find the items that should be hidden. It's less code and less indirection (IMHO) :-)

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeTreeGridItemView.java (hunk 11 above)
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeTreeGridItemView.java]: Fixed! ;-)

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeTreeGridItem.java (hunk 10 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeTreeGridItem.java]: Hmm.. you could do something like I show below.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeTreeGridItem.java [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+
+        final HTMLElement arrow = view.getArrow();
+
+        if (isCollapsed(arrow)) {
```
Comment (manstis) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeTreeGridItem.java]:
```
if(view.isCollapsed()) {
    view.expand();
} else {
    view.collapse();
}
```
and in the ```view```
```
public void expand() {
    CssHelper.asDownArrow(getArrow());
    presenter.expandSubDataTypes();
}
```
and in the ```presenter```:
```
public void expandSubDataTypes() {
    expandSubDataTypes(getDataType());
}
```

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeTreeGridItem.java (hunk 10 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeTreeGridItem.java]: I guess it's the leak of ```HTMLElement``` to the presenter I'm not keen on.

**Comment by jomarko** [comment]
@karreiro please address checkstyle violations:
```
[ERROR] src/main/java/org/kie/workbench/common/dmn/client/editors/types/common/DataTypeFactory.java:[29,8] (imports) UnusedImports: Unused import - com.google.gwt.core.client.GWT.
[ERROR] src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/common/JQueryTooltip.java:[8] (header) RegexpHeader: Line does not match expected header line of '^ \*     http:\/\/www\.apache\.org\/licenses\/LICENSE-2\.0$'.
[ERROR] src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/common/JQuerySelectPicker.java:[8] (header) RegexpHeader: Line does not match expected header line of '^ \*     http:\/\/www\.apache\.org\/licenses\/LICENSE-2\.0$'.

```

**Comment by jomarko** [comment]
@karreiro also some transitive dependencies found
```
[WARNING] Rule 0: de.is24.maven.enforcer.rules.IllegalTransitiveDependencyCheck failed with message:
Found 5 illegal transitive type dependencies in artifact 'org.kie.workbench:kie-wb-common-dmn-client:jar:7.10.0-SNAPSHOT':
1.) jsinterop.annotations.JsFunction
2.) jsinterop.annotations.JsMethod
3.) jsinterop.annotations.JsType
4.) jsinterop.base.JsPropertyMap
5.) jsinterop.base.JsPropertyMapOfAny
```

**Comment by jomarko** [comment]
Question about the **Cancel** button. it has different design in comparison with other cancel buttons in workbench. Is it intentional? I do not recall such detail from Liz's wireframes.

Data Type dialog
screenshot from 2018-08-03 08-57-29

Other dialogs
screenshot from 2018-08-03 08-57-43

**Comment by jomarko** [comment]
Question about attributes order, do we want to sort attributes (will file new ticket in that case) or do we want to follow precisely the order from the source file (as we do currently)?

**Comment by jomarko** [comment]
@karreiro I am done with manual review, still want to check code. But has to say I like this dialog a lot! Great job! Just some questions/comments about some corner cases in comments above and in the acceptance test in [DROOLS-2759](https://issues.jboss.org/browse/DROOLS-2759).

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeTreeGridItem.java [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+        HTMLElement getArrow();
+
+        void setupSelectComponent(final DataTypeSelect typeSelect);
+    }
+
+    static class CssHelper {
+
+        static void asRightArrow(final Element element) {
+            element.classList.add("fa-angle-right");
+            element.classList.remove("fa-angle-down");
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeTreeGridItem.java]: please introduce constants for "fa-angle-right" and "fa-angle-down"

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeSelectTest.java [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+
+        assertEquals(actual, expected);
+    }
+
+    @Test
+    public void testGetDefaultDataTypes() {
+
+        final List<DataType> expectedDataTypes = new ArrayList<>();
+
+        when(dataTypeUtils.defaultDataTypes()).thenReturn(expectedDataTypes);
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeSelectTest.java]: Could we please use non empty `expectedDataTypes`? Test in the current form will pass even if we change this line to `when(dataTypeUtils.defaultDataTypes()).thenReturn(new ArrayList<>());`

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeSelectTest.java [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+
+        assertEquals(actualDataTypes, expectedDataTypes);
+    }
+
+    @Test
+    public void testGetCustomDataTypes() {
+
+        final List<DataType> expectedDataTypes = new ArrayList<>();
+
+        when(dataTypeUtils.customDataTypes()).thenReturn(expectedDataTypes);
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeSelectTest.java]: Same comment as few lines above.

**Comment by jomarko** [comment]
And one friendly reminder with coverage increasement PR https://github.com/karreiro/kie-wb-common/pull/4

**Comment by karreiro** [comment]
@manstis I just applied the changes. Thank you for the suggestion.

---

@jomarko

> I) Question about the Cancel button. it has different design in comparison with other cancel buttons in workbench. Is it intentional? I do not recall such detail from Liz's wireframes.

I applied the standard style. However, probably it will be removed in a future PR (when I adequate the dialog to the new prototypes).


> II) Question about attributes order, do we want to sort attributes (will file a new ticket in that case) or do we want to follow precisely the order from the source file (as we do currently)?

I think that it's a good idea do sort attributes. So, yes, we can raise a JIRA for that.


> III) Regarding the comments that you added in the *DROOLS-2759*.

I think that we can raise a JIRA for the recursive scenario and the special chars scenario, however, let's ignore the deeply nested example for now. The dialog will suffer some changes with the new prototypes, and this issue will be addressed there.


> IV) `DataTypeSelectView.java:94` - Could you please teach me here a little bit? How it is possible to instantiate abstract class here, is it connected somehow to the fact it is native? Where comes its implementation from, please?

I'm just calling static methods there :-) (in fact, `JQueryTooltip` and `JQuerySelectPicker` are abstract to prevent the instantiation). You can check more about the JsInterop here: http://www.gwtproject.org/doc/latest/DevGuideCodingBasicsJsInterop.html in the "Importing a type from an external JavaScript script" section. I think that this page explains better than me :-)

> V) And one friendly reminder with coverage increasement PR -https://github.com/karreiro/kie-wb-common/pull/4/files
I checked the PR, and I think that I incorporate everything now, except the two topics that I commented above https://github.com/kiegroup/kie-wb-common/pull/2011#issuecomment-409683570.

Thank you.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeTreeGridView.java [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+        this.gridItems = gridItems;
+        this.attributeTooltip = attributeTooltip;
+        this.typeTooltip = typeTooltip;
+        this.translationService = translationService;
+    }
+
+    @PostConstruct
+    public void setupTooltips() {
+
+        final NodeList<Element> element = getElement().querySelectorAll("[data-toggle='tooltip']");
```
Comment (manstis) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeTreeGridView.java]: See...... you happily use a JQuery selector here, in this ```View``` but use it in the ```Presenter``` for ```DataTypeTreeGridItem``` above. It'd be nice to be consistent (and refactor ```DataTypeTreeGridItem``` to move some of the pure UI logic to the ```View```).

**Comment by jomarko** [comment]
@karreiro thanks for answers, I filed issues
- [DROOLS-2841](https://issues.jboss.org/browse/DROOLS-2841)
- [DROOLS-2842](https://issues.jboss.org/browse/DROOLS-2842)
- [DROOLS-2843](https://issues.jboss.org/browse/DROOLS-2843)

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeTreeGridTest.java [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+    public void testSetupItems() {
+
+        final DataType dataType = makeDataType("item", "iITem");
+        final List<DataTypeTreeGridItem> gridItems = new ArrayList<>();
+
+        doReturn(gridItems).when(treeGrid).makeTreeGridItems(dataType, 1);
+
+        treeGrid.setupItems(dataType);
+
+        verify(view).setupGridItems(gridItems);
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeTreeGridTest.java]: Missing `verify(treeGrid, never()).makeTreeGridItems(any(DataType.class), eq(2));` see https://github.com/karreiro/kie-wb-common/pull/4/commits/3918b7f5d8350586040b90e35b21b3f35af4df24

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeTreeGridItemView.java [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+    public void setDataType(final DataType dataType) {
+        setupRowMetadata(dataType);
+        setupArrow(dataType);
+        setupIndentationLevel();
+        setupDataTypeValues(dataType);
+    }
+
+    void setupRowMetadata(final DataType dataType) {
+
+        row.classList.add(dataType.isBasic() ? "basic-row" : "structure-row");
```
[the remark this answers is not in the document]
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeTreeGridItemView.java]: I prefer to keep these ones as they are. I agree that extracting variables and constants make the code more expressive and readable. However, sometimes extracting some strings creates an indirection for someplace distant (Uncle Bob talks about it in Clean Code when he discusses vertical distance). In this kind of scenario, I think that it's more readable to keep this values in this small context.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeTreeGridItem.java [hunk 19]
```diff
@@ -0,0 +1,156 @@
...
+    }
+
+    void collapseSubDataTypes() {
+        collapseSubDataTypes(getDataType());
+    }
+
+    void expandSubDataTypes(final DataType dataType) {
+        dataType.getSubDataTypes().forEach(subType -> {
+
+            final Element subDataTypeRow = getRowElement(subType);
```
[the remark this answers is not in the document]
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeTreeGridItem.java]: Fixed ;-)

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/TypePickerWidgetTest.java [hunk 20]
```diff
@@ -301,4 +307,17 @@ public void testEnable() {
...
+        final String value = "value";
+        final ClickEvent clickEvent = mock(ClickEvent.class);
+
+        when(typeSelector.getValue()).thenReturn(value);
+
+        picker.onClickTypeButton(clickEvent);
+
+        verify(dataTypeModal).show(value);
+    }
 }
```
[the remark this answers is not in the document]
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/TypePickerWidgetTest.java]: @jomarko I'm adding tests just for the changes in the `TypePickerWidget.java`:
changes

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeTreeGridTest.java (hunk 17 above)
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeTreeGridTest.java]: @jomarko Instead of asserting what is _not_ called. I'm verifying the amount of method calls (to ensure that we're verifying all of them). Like: 4 times in the `testSetupItemsWithSubItems` and 1 time in the `testSetupItemsWithoutSubItems`.

**Comment by karreiro** [comment]
@manstis @jomarko Changes applied. Thank you for the review ;-)

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeTreeGridItem.java [hunk 21]
```diff
@@ -0,0 +1,122 @@
...
+
+    public interface View extends UberElemental<DataTypeTreeGridItem> {
+
+        void setDataType(final DataType dataType);
+
+        void expand();
+
+        void collapse();
+
+        HTMLElement getArrow();
```
Comment (manstis) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeTreeGridItem.java]: Is this still needed?

Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/treegrid/DataTypeTreeGridItem.java]: Fixed.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/TypePickerWidgetTest.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (jomarko) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/types/TypePickerWidgetTest.java]: @karreiro thanks, understand, I did it [here](https://github.com/kiegroup/kie-wb-common/pull/2021), have a look if interested.
